File: bindings/java/umicp-core/src/main/java/com/hivellm/umicp/core/Matrix.java
package com.hivellm.umicp.core;

import org.jetbrains.annotations.NotNull;

/**
 * High-performance matrix and vector operations.
 *
 * <p>This class provides optimized operations for matrix and vector computations,
 * commonly used in machine learning and data processing pipelines. Operations include:
 * <ul>
 *   <li>Vector addition, subtraction, multiplication</li>
 *   <li>Dot product</li>
 *   <li>Cosine similarity</li>
 *   <li>Matrix multiplication</li>
 *   <li>Matrix transpose</li>
 *   <li>Vector normalization</li>
 * </ul>
 *
 * <p>Example usage:
 * <pre>{@code
 * try (Matrix matrix = new Matrix()) {
 *     // Dot product
 *     float[] a = {1.0f, 2.0f, 3.0f};
 *     float[] b = {4.0f, 5.0f, 6.0f};
 *     MatrixResult result = matrix.dotProduct(a, b);
 *     System.out.println("Dot product: " + result.getResult()); // 32.0
 *
 *     // Cosine similarity
 *     MatrixResult sim = matrix.cosineSimilarity(a, b);
 *     System.out.println("Similarity: " + sim.getSimilarity()); // 0.9746...
 *
 *     // Vector addition
 *     float[] sum = new float[3];
 *     matrix.vectorAdd(a, b, sum);
 *     System.out.println("Sum: " + Arrays.toString(sum)); // [5.0, 7.0, 9.0]
 * }
 * }</pre>
 *
 * <p>This is a pure Java implementation. For better performance, consider using
 * the JNI-backed version which leverages native SIMD acceleration.
 *
 * @author HiveLLM Team
 * @version 1.0.0
 * @since 1.0.0
 * @see MatrixResult
 */
public class Matrix implements AutoCloseable {

    private static final double EPSILON = 1e-10;

    /**
     * Creates a new Matrix instance.
     */
    public Matrix() {
        // Pure Java implementation - no native resources
    }

    /**
     * Adds two vectors element-wise: result = a + b
     *
     * @param a the first vector
     * @param b the second vector
     * @param result the output vector
     * @return the operation result
     */
    @NotNull
    public MatrixResult vectorAdd(@NotNull float[] a, @NotNull float[] b, @NotNull float[] result) {
        if (a.length != b.length || a.length != result.length) {
            return MatrixResult.failure("Vector dimensions must match");
        }

        try {
            for (int i = 0; i < a.length; i++) {
                result[i] = a[i] + b[i];
            }
            return MatrixResult.success();
        } catch (Exception e) {
            return MatrixResult.failure("Vector addition failed: " + e.getMessage());
        }
    }

    /**
     * Subtracts two vectors element-wise: result = a - b
     *
     * @param a the first vector
     * @param b the second vector
     * @param result the output vector
     * @return the operation result
     */
    @NotNull
    public MatrixResult vectorSubtract(@NotNull float[] a, @NotNull float[] b, @NotNull float[] result) {
        if (a.length != b.length || a.length != result.length) {
            return MatrixResult.failure("Vector dimensions must match");
        }

        try {
            for (int i = 0; i < a.length; i++) {
                result[i] = a[i] - b[i];
            }
            return MatrixResult.success();
        } catch (Exception e) {
            return MatrixResult.failure("Vector subtraction failed: " + e.getMessage());
        }
    }

    /**
     * Multiplies two vectors element-wise: result = a * b
     *
     * @param a the first vector
     * @param b the second vector
     * @param result the output vector
     * @return the operation result
     */
    @NotNull
    public MatrixResult vectorMultiply(@NotNull float[] a, @NotNull float[] b, @NotNull float[] result) {
        if (a.length != b.length || a.length != result.length) {
            return MatrixResult.failure("Vector dimensions must match");
        }

        try {
            for (int i = 0; i < a.length; i++) {
                result[i] = a[i] * b[i];
            }
            return MatrixResult.success();
        } catch (Exception e) {
            return MatrixResult.failure("Vector multiplication failed: " + e.getMessage());
        }
    }

    /**
     * Computes the dot product of two vectors: a · b
     *
     * @param a the first vector
     * @param b the second vector
     * @return the operation result containing the dot product
     */
    @NotNull
    public MatrixResult dotProduct(@NotNull float[] a, @NotNull float[] b) {
        if (a.length != b.length) {
            return MatrixResult.failure("Vector dimensions must match");
        }

        try {
            double sum = 0.0;
            for (int i = 0; i < a.length; i++) {
                sum += a[i] * b[i];
            }
            return MatrixResult.success(sum);
        } catch (Exception e) {
            return MatrixResult.failure("Dot product failed: " + e.getMessage());
        }
    }

    /**
     * Computes the cosine similarity between two vectors: (a · b) / (||a|| * ||b||)
     *
     * <p>The result is in the range [-1, 1], where:
     * <ul>
     *   <li>1 indicates identical direction</li>
     *   <li>0 indicates orthogonal vectors</li>
     *   <li>-1 indicates opposite direction</li>
     * </ul>
     *
     * @param a the first vector
     * @param b the second vector
     * @return the operation result containing the similarity score
     */
    @NotNull
    public MatrixResult cosineSimilarity(@NotNull float[] a, @NotNull float[] b) {
        if (a.length != b.length) {
            return MatrixResult.failure("Vector dimensions must match");
        }

        try {
            double dotProduct = 0.0;
            double normA = 0.0;
            double normB = 0.0;

            for (int i = 0; i < a.length; i++) {
                dotProduct += a[i] * b[i];
                normA += a[i] * a[i];
                normB += b[i] * b[i];
            }

            normA = Math.sqrt(normA);
            normB = Math.sqrt(normB);

            if (normA < EPSILON || normB < EPSILON) {
                return MatrixResult.failure("Zero-magnitude vector encountered");
            }

            double similarity = dotProduct / (normA * normB);
            return MatrixResult.successWithSimilarity(similarity);
        } catch (Exception e) {
            return MatrixResult.failure("Cosine similarity failed: " + e.getMessage());
        }
    }

    /**
     * Normalizes a vector to unit length (L2 normalization).
     *
     * @param vector the vector to normalize (modified in-place)
     * @return the operation result
     */
    @NotNull
    public MatrixResult normalize(@NotNull float[] vector) {
        try {
            double norm = 0.0;
            for (float v : vector) {
                norm += v * v;
            }
            norm = Math.sqrt(norm);

            if (norm < EPSILON) {
                return MatrixResult.failure("Cannot normalize zero-magnitude vector");
            }

            for (int i = 0; i < vector.length; i++) {
                vector[i] = (float) (vector[i] / norm);
            }

            return MatrixResult.success();
        } catch (Exception e) {
            return MatrixResult.failure("Normalization failed: " + e.getMessage());
        }
    }

    /**
     * Multiplies two matrices: result = A * B
     *
     * <p>Matrix A is m×n, matrix B is n×p, result is m×p.
     *
     * @param a the first matrix (row-major order)
     * @param b the second matrix (row-major order)
     * @param result the output matrix (row-major order)
     * @param m number of rows in A
     * @param n number of columns in A (rows in B)
     * @param p number of columns in B
     * @return the operation result
     */
    @NotNull
    public MatrixResult matrixMultiply(@NotNull float[] a, @NotNull float[] b,
                                      @NotNull float[] result, int m, int n, int p) {
        if (a.length != m * n) {
            return MatrixResult.failure("Matrix A dimensions don't match: expected " + (m*n) + ", got " + a.length);
        }
        if (b.length != n * p) {
            return MatrixResult.failure("Matrix B dimensions don't match: expected " + (n*p) + ", got " + b.length);
        }
        if (result.length != m * p) {
            return MatrixResult.failure("Result dimensions don't match: expected " + (m*p) + ", got " + result.length);
        }

        try {
            // Initialize result to zero
            for (int i = 0; i < result.length; i++) {
                result[i] = 0.0f;
            }

            // Perform matrix multiplication
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < p; j++) {
                    for (int k = 0; k < n; k++) {
                        result[i * p + j] += a[i * n + k] * b[k * p + j];
                    }
                }
            }

            return MatrixResult.success();
        } catch (Exception e) {
            return MatrixResult.failure("Matrix multiplication failed: " + e.getMessage());
        }
    }

    /**
     * Transposes a matrix: result = A^T
     *
     * @param input the input matrix (row-major order)
     * @param output the output matrix (row-major order)
     * @param rows number of rows in input
     * @param cols number of columns in input
     * @return the operation result
     */
    @NotNull
    public MatrixResult transpose(@NotNull float[] input, @NotNull float[] output,
                                  int rows, int cols) {
        if (input.length != rows * cols) {
            return MatrixResult.failure("Input dimensions don't match: expected " + (rows*cols) + ", got " + input.length);
        }
        if (output.length != rows * cols) {
            return MatrixResult.failure("Output dimensions don't match: expected " + (rows*cols) + ", got " + output.length);
        }

        try {
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    output[j * rows + i] = input[i * cols + j];
                }
            }
            return MatrixResult.success();
        } catch (Exception e) {
            return MatrixResult.failure("Matrix transpose failed: " + e.getMessage());
        }
    }

    /**
     * Computes the magnitude (L2 norm) of a vector.
     *
     * @param vector the input vector
     * @return the operation result containing the magnitude
     */
    @NotNull
    public MatrixResult magnitude(@NotNull float[] vector) {
        try {
            double sum = 0.0;
            for (float v : vector) {
                sum += v * v;
            }
            return MatrixResult.success(Math.sqrt(sum));
        } catch (Exception e) {
            return MatrixResult.failure("Magnitude calculation failed: " + e.getMessage());
        }
    }

    /**
     * Closes this matrix instance and releases any resources.
     *
     * <p>In the pure Java implementation, this method does nothing.
     * It's provided for compatibility with JNI-based implementations.
     */
    @Override
    public void close() {
        // No resources to release in pure Java implementation
    }
}

File: bindings/kotlin/src/main/kotlin/com/hivellm/umicp/core/Matrix.kt
package com.hivellm.umicp.core

import com.hivellm.umicp.types.MatrixException
import kotlin.math.sqrt

/**
 * Result wrapper for matrix operations
 */
sealed class MatrixResult<out T> {
    data class Success<T>(val value: T) : MatrixResult<T>()
    data class Failure(val error: String) : MatrixResult<Nothing>()

    val isSuccess: Boolean get() = this is Success
    val isFailure: Boolean get() = this is Failure

    fun getOrNull(): T? = when (this) {
        is Success -> value
        is Failure -> null
    }

    fun getOrThrow(): T = when (this) {
        is Success -> value
        is Failure -> throw MatrixException(error)
    }

    inline fun <R> map(transform: (T) -> R): MatrixResult<R> = when (this) {
        is Success -> Success(transform(value))
        is Failure -> this
    }

    inline fun onSuccess(action: (T) -> Unit): MatrixResult<T> {
        if (this is Success) action(value)
        return this
    }

    inline fun onFailure(action: (String) -> Unit): MatrixResult<T> {
        if (this is Failure) action(error)
        return this
    }
}

/**
 * Matrix operations for UMICP
 */
object Matrix {
    /**
     * Compute dot product of two vectors
     *
     * @param a First vector
     * @param b Second vector
     * @return Dot product result
     */
    fun dotProduct(a: FloatArray, b: FloatArray): MatrixResult<Float> {
        if (a.size != b.size) {
            return MatrixResult.Failure("Vector dimensions must match: ${a.size} != ${b.size}")
        }

        var sum = 0f
        for (i in a.indices) {
            sum += a[i] * b[i]
        }

        return MatrixResult.Success(sum)
    }

    /**
     * Compute cosine similarity between two vectors
     *
     * @param a First vector
     * @param b Second vector
     * @return Cosine similarity in range [-1, 1]
     */
    fun cosineSimilarity(a: FloatArray, b: FloatArray): MatrixResult<Float> {
        if (a.size != b.size) {
            return MatrixResult.Failure("Vector dimensions must match: ${a.size} != ${b.size}")
        }

        var dotProd = 0f
        var magA = 0f
        var magB = 0f

        for (i in a.indices) {
            dotProd += a[i] * b[i]
            magA += a[i] * a[i]
            magB += b[i] * b[i]
        }

        magA = sqrt(magA)
        magB = sqrt(magB)

        if (magA == 0f || magB == 0f) {
            return MatrixResult.Failure("Cannot compute similarity with zero magnitude vector")
        }

        return MatrixResult.Success(dotProd / (magA * magB))
    }

    /**
     * Compute magnitude (L2 norm) of a vector
     *
     * @param vector Input vector
     * @return Magnitude
     */
    fun magnitude(vector: FloatArray): MatrixResult<Float> {
        var sum = 0f
        for (value in vector) {
            sum += value * value
        }
        return MatrixResult.Success(sqrt(sum))
    }

    /**
     * Normalize a vector to unit length
     *
     * @param vector Input vector
     * @return Normalized vector
     */
    fun normalize(vector: FloatArray): MatrixResult<FloatArray> {
        val mag = magnitude(vector).getOrThrow()

        if (mag == 0f) {
            return MatrixResult.Failure("Cannot normalize zero vector")
        }

        val normalized = FloatArray(vector.size) { i -> vector[i] / mag }
        return MatrixResult.Success(normalized)
    }

    /**
     * Add two vectors element-wise
     *
     * @param a First vector
     * @param b Second vector
     * @return Result vector
     */
    fun vectorAdd(a: FloatArray, b: FloatArray): MatrixResult<FloatArray> {
        if (a.size != b.size) {
            return MatrixResult.Failure("Vector dimensions must match: ${a.size} != ${b.size}")
        }

        val result = FloatArray(a.size) { i -> a[i] + b[i] }
        return MatrixResult.Success(result)
    }

    /**
     * Subtract two vectors element-wise
     *
     * @param a First vector
     * @param b Second vector
     * @return Result vector
     */
    fun vectorSubtract(a: FloatArray, b: FloatArray): MatrixResult<FloatArray> {
        if (a.size != b.size) {
            return MatrixResult.Failure("Vector dimensions must match: ${a.size} != ${b.size}")
        }

        val result = FloatArray(a.size) { i -> a[i] - b[i] }
        return MatrixResult.Success(result)
    }

    /**
     * Multiply two vectors element-wise (Hadamard product)
     *
     * @param a First vector
     * @param b Second vector
     * @return Result vector
     */
    fun vectorMultiply(a: FloatArray, b: FloatArray): MatrixResult<FloatArray> {
        if (a.size != b.size) {
            return MatrixResult.Failure("Vector dimensions must match: ${a.size} != ${b.size}")
        }

        val result = FloatArray(a.size) { i -> a[i] * b[i] }
        return MatrixResult.Success(result)
    }

    /**
     * Matrix multiplication: C = A × B
     *
     * @param a Matrix A (m × n)
     * @param b Matrix B (n × p)
     * @param m Number of rows in A
     * @param n Number of columns in A (rows in B)
     * @param p Number of columns in B
     * @return Result matrix C (m × p)
     */
    fun matrixMultiply(
        a: FloatArray,
        b: FloatArray,
        m: Int,
        n: Int,
        p: Int
    ): MatrixResult<FloatArray> {
        if (a.size != m * n) {
            return MatrixResult.Failure("Matrix A size mismatch: expected ${m * n}, got ${a.size}")
        }
        if (b.size != n * p) {
            return MatrixResult.Failure("Matrix B size mismatch: expected ${n * p}, got ${b.size}")
        }

        val result = FloatArray(m * p)

        for (i in 0 until m) {
            for (j in 0 until p) {
                var sum = 0f
                for (k in 0 until n) {
                    sum += a[i * n + k] * b[k * p + j]
                }
                result[i * p + j] = sum
            }
        }

        return MatrixResult.Success(result)
    }

    /**
     * Transpose a matrix
     *
     * @param matrix Input matrix (rows × cols)
     * @param rows Number of rows
     * @param cols Number of columns
     * @return Transposed matrix (cols × rows)
     */
    fun transpose(matrix: FloatArray, rows: Int, cols: Int): MatrixResult<FloatArray> {
        if (matrix.size != rows * cols) {
            return MatrixResult.Failure("Matrix size mismatch: expected ${rows * cols}, got ${matrix.size}")
        }

        val result = FloatArray(rows * cols)

        for (i in 0 until rows) {
            for (j in 0 until cols) {
                result[j * rows + i] = matrix[i * cols + j]
            }
        }

        return MatrixResult.Success(result)
    }
}

/**
 * Extension functions for convenient matrix operations
 */

/**
 * Compute dot product with another vector
 */
infix fun FloatArray.dot(other: FloatArray): Float =
    Matrix.dotProduct(this, other).getOrThrow()

/**
 * Compute cosine similarity with another vector
 */
infix fun FloatArray.cosineSim(other: FloatArray): Float =
    Matrix.cosineSimilarity(this, other).getOrThrow()

/**
 * Compute magnitude of vector
 */
fun FloatArray.magnitude(): Float =
    Matrix.magnitude(this).getOrThrow()

/**
 * Normalize vector to unit length
 */
fun FloatArray.normalize(): FloatArray =
    Matrix.normalize(this).getOrThrow()

/**
 * Add two vectors
 */
operator fun FloatArray.plus(other: FloatArray): FloatArray =
    Matrix.vectorAdd(this, other).getOrThrow()

/**
 * Subtract two vectors
 */
operator fun FloatArray.minus(other: FloatArray): FloatArray =
    Matrix.vectorSubtract(this, other).getOrThrow()

/**
 * Element-wise multiplication
 */
operator fun FloatArray.times(other: FloatArray): FloatArray =
    Matrix.vectorMultiply(this, other).getOrThrow()

